### PR TITLE
Adding ForceArgs module

### PR DESCRIPTION
**Pull request by AlexeyKurakin**
This address issue #37

### CODE REVIEW

File: objax/module.py [hunk 1]
```diff
@@ -53,6 +53,54 @@ def __call__(self, *args, **kwargs):
...
+                        v = v._wrapped
+                        module[idx] = v
+                    ForceArgs.undo(v)
+        else:
+            for k, v in module.__dict__.items():
+                if isinstance(v, Module):
+                    if isinstance(v, ForceArgs):
+                        v = v._wrapped
+                        module.__dict__[k] = v
+                    ForceArgs.undo(v)
```
Comment (david-berthelot) [inline objax/module.py:81]: The code can be greatly simplified noting that:
1. `ForceArgs` called on a root of module tree, returns a new module.
2. Consequently calling undo on this root should also return a module, the original one.

Also I think `setattr` is the preferred way to write attributes in Python, so I also made that change.

```suggestion
    def undo(module: Module):
        """Undo ForceArgs on each submodule of the module.

        Args:
            module: module for which to undo ForceArgs.
        """
        if isinstance(module, ForceArgs):
            return ForceArgs.undo(module.__wrapped__)

        if isinstance(module, ModuleList):
            for idx, v in enumerate(module):
                if isinstance(v, Module):
                    module[idx] = ForceArgs.undo(v)
            return module

        for k, v in module.__dict__.items():
            if isinstance(v, Module):
                setattr(module, k, ForceArgs.undo(v))
        return module
```

File: objax/module.py [hunk 2]
```diff
@@ -53,6 +53,54 @@ def __call__(self, *args, **kwargs):
...
+        """Initializes ForceArgs.
+
+        Args:
+            module: base module, which argument will be overridden.
+            kwargs: values of keyword arguments which will be forced to use.
+        """
+        if isinstance(module, ForceArgs):
+            raise ValueError('Can not force arguments on ForceArgs module.')
+        self._wrapped = module
+        self.forced_kwargs = kwargs
```
Comment (david-berthelot) [inline objax/module.py:93]: Two things:
1. `base_module` can be simply called `module`.
2. Rename `_wrapped` to `__wrapped__` which is used by python to determine the signature of a function.
```suggestion
    def __init__(self, module: Module, **kwargs):
        """Initializes ForceArgs.

        Args:
            module: module which arguments will be overridden.
            kwargs: values of keyword arguments which will be forced to use.
        """
        if isinstance(module, ForceArgs):
            raise ValueError('Can not force arguments on ForceArgs module.')
        self.__wrapped__ = module
        self.forced_kwargs = kwargs
```

File: objax/module.py [hunk 3]
```diff
@@ -53,6 +53,54 @@ def __call__(self, *args, **kwargs):
...
+
+    def vars(self, scope: str = '') -> VarCollection:
+        return self._wrapped.vars(scope=scope)
```
Comment (david-berthelot) [inline objax/module.py:96]:
```suggestion
        return self.__wrapped__.vars(scope=scope)
```

File: objax/module.py [hunk 4]
```diff
@@ -53,6 +53,54 @@ def __call__(self, *args, **kwargs):
...
+
+    def __call__(self, *args, **kwargs):
+        kwargs.update(self.forced_kwargs)
+        args, kwargs = merge_args_kwargs(args, kwargs, self._wrapped)
+        return self._wrapped(*args, **kwargs)
```
Comment (david-berthelot) [inline objax/module.py:101]: The current method had some issue with Python, I explain it in details on the suggestion for `override_args_kwargs`.
```suggestion
        args, kwargs = override_args_kwargs(args, kwargs, self.forced_kwargs, self.__wrapped__)
        return self.__wrapped__(*args, **kwargs)
```

File: objax/module.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
 from jax.interpreters.pxla import ShardedDeviceArray
 
 from objax.typing import JaxArray
-from objax.util import positional_args_names
+from objax.util import merge_args_kwargs, positional_args_names
```
Comment (david-berthelot) [inline objax/module.py:25]: See code comments below to see why I changed the function (it's not just a rename).
```suggestion
from objax.util import overridee_args_kwargs, positional_args_names
```

File: objax/util/util.py [hunk 6]
```diff
@@ -13,7 +13,7 @@
 # limitations under the License.
 
 __all__ = ['EasyDict', 'args_indexes', 'dummy_context_mgr', 'ilog2', 'local_kwargs', 'map_to_device',
-           'multi_host_barrier', 'positional_args_names', 'to_tuple']
+           'merge_args_kwargs', 'multi_host_barrier', 'positional_args_names', 'to_tuple']
```
Comment (david-berthelot) [inline objax/util/util.py:16]:
```suggestion
           'override_args_kwargs', 'multi_host_barrier', 'positional_args_names', 'to_tuple']
```

File: objax/util/util.py [hunk 7]
```diff
@@ -69,6 +69,32 @@ def local_kwargs(kwargs: dict, f: Callable) -> dict:
...
+    """
+    args = list(args)
+    kwargs = local_kwargs(kwargs, f)
+    p = inspect.signature(f).parameters
+    for idx, (k, v) in enumerate(p.items()):
+        if (idx >= len(args)) or (v.kind == inspect.Parameter.VAR_POSITIONAL):
+            break
+        if k in kwargs:
+            args[idx] = kwargs.pop(k)
+    return args, kwargs
```
Comment (david-berthelot) [inline objax/util/util.py:95]: I propose a significant change here since the original implementation was hiding code bugs.

Here's an example:
```python
class M(Module):
    def __call__(self, a, b, c):
        return 0

g = objax.ForceArgs(M(), b=1)
g(1, 2, 3, b=4) # should error but didn't in the original code.
f(1, 2, 3, b=4) # the ForceArgs version should error just like the original function f.
```

There were other failure scenarios. But the gist is that when passing improper/invalid argument, the behavior of a `ForceArgs(f)` should be the one of `f`. Consequently, I made the following suggestion to only override the value of an argument if present without modifying arguments in any other way.

```suggestion
def override_args_kwargs(args: Tuple[Any, ...],
                         kwargs: Dict[str, Any],
                         override:  Dict[str, Any],
                         f: Callable) -> Tuple[List, Dict[str, Any]]:
    """Override positional and keyword arguments according to signature of the function.

    Replace the values in args and kwargs with the corresponding override entries.

    Args:
        args: positional arguments.
        kwargs: keyword arguments.
        override: arguments for which to override the value.
        f: callable, for which arguments should be fixed.

    Return:
        args: updated list of positional arguments.
        kwargs: updated dictionary of keyword arguments.
    """
    args = list(args)
    params = inspect.signature(f).parameters
    override = override.copy()
    for i, (k, v) in enumerate(itertools.islice(params.items(), len(args))):
        if v.kind not in (inspect.Parameter.POSITIONAL_ONLY, inspect.Parameter.POSITIONAL_OR_KEYWORD):
            break
        if k in override:
            args[i] = override.pop(k)
    return args, {**kwargs, **override}
```

File: tests/module.py [hunk 8]
```diff
@@ -75,6 +107,24 @@ def test_module_wrapper(self):
...
+        np.testing.assert_almost_equal(model(x, some_arg=1.0), [6., 7.])
+        # set forced args in a list
+        model.block1.seq[0] = objax.ForceArgs(model.block1.seq[0], some_arg=-1.0)
+        np.testing.assert_almost_equal(model(x, some_arg=0.0), [-1., 0.])
+        np.testing.assert_almost_equal(model(x, some_arg=1.0), [4., 5.])
+        # undo force args
+        objax.ForceArgs.undo(model)
+        np.testing.assert_almost_equal(model(x, some_arg=0.0), [1., 2.])
+        np.testing.assert_almost_equal(model(x, some_arg=1.0), [8., 9.])
+
```
Comment (david-berthelot) [inline tests/module.py:127]: Add tests for negative cases with `self.assertRaises` to make sure a `ForceArgs(f)` produces the same errors as `f` when called improperly (there was one example in a previous suggestion).

**Review by david-berthelot — CHANGES_REQUESTED** [review]

File: objax/module.py (hunk 2 above)
Comment (AlexeyKurakin) [reply objax/module.py:93]: Done

File: objax/module.py (hunk 4 above)
Comment (AlexeyKurakin) [reply objax/module.py:101]: Done

File: objax/util/util.py (hunk 7 above)
Comment (AlexeyKurakin) [reply objax/util/util.py:95]: Done

File: objax/module.py (hunk 1 above)
Comment (AlexeyKurakin) [reply objax/module.py:81]: I think user should not be allowed to call `ForceArgs.undo` on instance of `ForceArgs` because it may lead to surprising behavior from user point of view:

```python
# override args of one operation in the model
m = SomeModel()
m.bn1 = ForceArgs(m.bn1, training=False)
y = m(x, training=True)  # argument of m.bn1 is overriden

ForceArgs.undo(m)  # reset all overrides

# override args of entire model
pred_m = ForceArgs(m, training=False)
y_pred = pred_m(x)  # training is overriden for entire model

# if ForceArgs.undo can be called on instance of ForceArgs following line will do nothing
# however user may expect that arguments override will be undone
ForceArgs.undo(pred_m)
```

There is a separate question whether `ForceArgs.undo(m)` should return `m`.
I think it's ok to add, but not sure how useful. I just allow users to write code like `y = ForceArgs.undo(m)(x)` which might be actually somewhat confusing

**Comment by AlexeyKurakin** [comment]
@david-berthelot code updated, please take a look

File: objax/module.py (hunk 1 above)
Comment (david-berthelot) [reply objax/module.py]: 1. One can call `ForceArgs` on the root, therefore calling `ForceArgs.undo` should be doable on the root too.
2. Returning a module results in simpler implementation too.

**Review by david-berthelot — CHANGES_REQUESTED** [review]
The `undo` method looks complicated and doesn't work on the root.

**Comment by AlexeyKurakin** [comment]
@david-berthelot ForceArgs updated. Please take a look

File: objax/module.py [hunk 9]
```diff
@@ -53,6 +54,68 @@ def __call__(self, *args, **kwargs):
...
 
 
+class ForceArgs(Module):
+    """Forces override of arguments of given module."""
+
+    ANY = namedtuple('Any', ())
+    """Token used in `ForceArgs.undo` to indicate undo of all values of specific argument."""
+
+    @staticmethod
+    def _remove_args(force_args, args_to_remove: dict):
```
Comment (david-berthelot) [inline objax/module.py:64]:
```suggestion
    def _remove_args(force_args: 'ForceArgs', args_to_remove: Dict[str, Any]):
```

File: objax/module.py [hunk 10]
```diff
@@ -53,6 +54,68 @@ def __call__(self, *args, **kwargs):
...
+        if isinstance(module, ForceArgs):
+            ForceArgs._remove_args(module, kwargs)
+            ForceArgs.undo(module.__wrapped__, **kwargs)
+        elif isinstance(module, ModuleList):
+            for idx, v in enumerate(module):
+                if isinstance(v, Module):
+                    ForceArgs.undo(v, **kwargs)
+                    if isinstance(v, ForceArgs) and not v.forced_kwargs:
+                        v = v.__wrapped__
+                        module[idx] = v
```
Comment (david-berthelot) [inline objax/module.py:92]:
```suggestion
                        module[idx] = v.__wrapped__
```

File: objax/module.py [hunk 11]
```diff
@@ -53,6 +54,68 @@ def __call__(self, *args, **kwargs):
...
+        else:
+            for k, v in module.__dict__.items():
+                if isinstance(v, Module):
+                    ForceArgs.undo(v, **kwargs)
+                    if isinstance(v, ForceArgs) and not v.forced_kwargs:
+                        v = v.__wrapped__
+                        setattr(module, k, v)
```
Comment (david-berthelot) [inline objax/module.py:99]:
```suggestion
                        setattr(module, k, v.__wrapped__)
```

**Review by david-berthelot — CHANGES_REQUESTED** [review]
Looks very good, only suggested some minor typing hints and code simplification.

File: objax/module.py [hunk 12]
```diff
@@ -53,6 +54,66 @@ def __call__(self, *args, **kwargs):
...
 
 
+class ForceArgs(Module):
+    """Forces override of arguments of given module."""
+
+    ANY = namedtuple('Any', ())
+    """Token used in `ForceArgs.undo` to indicate undo of all values of specific argument."""
+
+    @staticmethod
+    def _remove_args(force_args: 'ForceArgs', args_to_remove: Dict[str, Any]):
```
Comment (david-berthelot) [inline objax/module.py:63]: I just realized this should probably not be a static method.
```suggestion
    def remove_args(self, args_to_remove: Dict[str, Any]):
```

File: objax/module.py [hunk 13]
```diff
@@ -53,6 +54,66 @@ def __call__(self, *args, **kwargs):
...
+        """Removes given arguments from override list of `ForceArgs` instance."""
+        if not args_to_remove:
+            force_args.forced_kwargs = {}
+        else:
+            force_args.forced_kwargs = {k: v for k, v in force_args.forced_kwargs.items()
+                                        if (k not in args_to_remove) or (args_to_remove[k] not in (v, ForceArgs.ANY))}
```
Comment (david-berthelot) [inline objax/module.py:69]:
```suggestion
            self.forced_kwargs = {}
        else:
            self.forced_kwargs = {k: v for k, v in self.forced_kwargs.items()
                                        if (k not in args_to_remove) or (args_to_remove[k] not in (v, ForceArgs.ANY))}
```

File: objax/module.py [hunk 14]
```diff
@@ -53,6 +54,66 @@ def __call__(self, *args, **kwargs):
...
+
+        Args:
+            module: module for which to undo ForceArgs.
+            **kwargs: dictionary of argument overrides to undo.
+                `name=val` remove override for value `val` of argument `name`.
+                `name=ForceArgs.ANY` remove all overrides of argument `name`.
+                If `**kwargs` is empty then all overrides will be undone.
+        """
+        if isinstance(module, ForceArgs):
+            ForceArgs._remove_args(module, kwargs)
```
Comment (david-berthelot) [inline objax/module.py:83]:
```suggestion
            module.remove_args(kwargs)
```

**Review by david-berthelot — CHANGES_REQUESTED** [review]
More suggestions

**Review by david-berthelot — APPROVED** [review]
